Sprint 3/step 1

piladik opened the pull request.

piladik committed:
Sprint 3 initial commit
1. Добавил папку pages и создал соответствующие ей файлы
2. Добавил маршрутизацию до созданых страниц
3. Теперь активная страница подсвечивается в header'e

piladik committed:
Удалил ненужные файлы

piladik committed:
Разбил profile-page на компоненты

piladik committed:
Перенес вызов модального окна из ingredients-card в burger-ingredients

piladik committed:
Реализовал указаное в ТЗ поведение модального окна. При клике на ингредиент открывается модальное окно, но при переходе по ссылке открывается страница ингредиента

piladik committed:
Заменил ссылки в header на компонент NavLink

piladik committed:
API features
1. Добавил экшены и редьюсеры для аутентификации и авторизации
2. Добавил возможность регистрации и авторизации через /register и /login
3. Добавил проверку токенов при рендере компонента App:
    -теперь при рендере компонента App происходит диспатч акшена getUser, который делает запрос к /auth/user при наличии accessToken.
    -при отсутствии accessToken выполняется запрос к /auth/token и используя новый accessToken выполняется запрос к /auth/user
    -экшн getUser также сохраняет данный о пользователе в redux
4. Если пользователь авторизован, то в хедере вместо "Личный кабинет" отображается его имя.

piladik committed:
Authorization update:
1. При успешной авторизации и регистрации происходит переадресация на главную страницу
2. Доработаны кнопки навигации на странице профиля:
    -активная кнопка теперь подсвечивается
    -при нажатии на выход диспачится экшн logout

piladik committed:
Password reset added

piladik committed:
Добавлена возможность обновлять данные аккаунта во вкладке личный кабинет

piladik committed:
Profile Navigation Update
1. Заменил компонент ProfilePage на компонент ProfileSharedLayout:
    -теперь в компоненте ProfileSharedLayout есть навигационная панель, с помощью которой можно переключаться между табами
    -Добавил компонент ProfileOrders
    -компоненты ProfileInfo и ProfileOrders теперь рендерятся через <Outlet>
2. Пофиксил баг из-за которого accessToken не обновлялся

piladik committed:
ProtecterRouting implemented
1. Добавил защищенные маршруты
2. Добавил Preloader

piladik committed:
Добавил защиту для роута /reset-password

piladik committed:
Делать заказы теперь могуть только авторизованные пользователи

piladik committed:
Добавил проверку PropTypes

piladik committed:
Повесил на все формы слушатель onSubmit и убрал с кнопок onClick

ArslanDev666 commented on src/components/preloader/preloader.jsx:
```diff
@@ -0,0 +1,21 @@
+        color="#4ffae9"
+        ariaLabel="tail-spin-loading"
+        radius="1"
+        wrapperStyle={{}}
+        wrapperClass=""
```
**Надо исправить:**

Думаю пропсы `wrapperStyle` и `wrapperClass`, лучше вообшще не указывать.

ArslanDev666 commented on src/utils/burger-api.js:
```diff
@@ -24,3 +45,97 @@ export const postOrderRequest = async (ingredientsId) => {
+  });
+};
+
+export const updateUserRequest = async ({ name, email, password = null }) => {
+  if (password === null) {
+    return await request(`${BURGER_BASE_API}/auth/user`, {
+      method: "PATCH",
+      headers: {
+        "Content-Type": "application/json",
+        Authorization: getCookie("accessToken"),
+      },
+      body: JSON.stringify({ name: name, email: email }),
+    });
+  } else {
+    return await request(`${BURGER_BASE_API}/auth/user`, {
+      method: "PATCH",
+      headers: {
+        "Content-Type": "application/json",
+        Authorization: getCookie("accessToken"),
+      },
+      body: JSON.stringify({ name: name, email: email, password: password }),
+    });
+  }
+};
```
**Можно лучше:**

я думаю, тут можно не делать такое условие, попробуйте просто создать объект, и если нужно к нему добавлять `password`

ArslanDev666 commented on src/utils/burger-api.js:
```diff
@@ -24,3 +45,97 @@ export const postOrderRequest = async (ingredientsId) => {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+      Authorization: getCookie("accessToken"),
```
**Можно лучше:**

я могу ошибаться, но вроде бы тут не должен передаваться токен))

ArslanDev666 commented on src/utils/burger-api.js:
```diff
@@ -24,3 +45,97 @@ export const postOrderRequest = async (ingredientsId) => {
+  });
+};
+
+export const updateUserRequest = async ({ name, email, password = null }) => {
```
**Надо исправить:**

разве в этом методе, вам не нужно использовать `requestWithRefresh`?

ArslanDev666 commented on src/services/reducers/auth.js:
```diff
@@ -0,0 +1,166 @@
+} from "../actions/auth";
+
+const initialState = {
+  user: {},
```
**Надо исправить:**

Если значения нет, то нужно использовать null, а не пустой объект.

ArslanDev666 commented on src/pages/register-page.jsx:
```diff
@@ -0,0 +1,22 @@
+export function RegisterPage() {
+  const { isLoggedIn } = useSelector((store) => store.auth);
+  const navigate = useNavigate();
+  const [form, setForm] = useState({ email: "", password: "", name: "" });
```
**Можно лучше:**

Вы ведь не используйте `form` и `setForm` в компоненте `RegisterPage`, почему бы, не обьявить их в `RegisterForm`. 




**Можно лучше**

Можно сделать универсальный кастомный хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues) {
  const [values, setValues] = useState(inputValues);

  const handleChange = event => {
    const { value, name } = event.target;
    setValues({ ...values, [name]: value });
  };
  return { values, handleChange, setValues };
}
```

Этот код помещают в отдельный файл useForm.js в папке hooks и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```js
const { values, handleChange, setValues } = useForm({...});
```

ArslanDev666 commented on src/components/register-form/register-form.jsx:
```diff
@@ -0,0 +1,89 @@
+    dispatch(register(form));
+  };
+  return (
+    <form onSubmit={onSubmit}>
```
**Вы молодец:**

Вы правильно повесели обработчик событий на форму)) многие по ошибке вешают обработчик на событие `onClick` кнопки.

ArslanMustafin reviewed (changes requested):
## Code Review

Здравствуйте!

Отличная работа.

### **Вы молодец!**

- Сборка и запуск проекта выполняются без ошибок;
- Теперь у вас есть роутинг;
- Есть регистрация и авторизация;

### **Но есть несколько замечаний, которые нужно исправить:**

- В запросе `/orders` нужно чтобы id у булок были в начале и в конце массива: <https://disk.yandex.com/i/jxtQneVyvCAmYg>;

- Вы должны добавлять в хедер запроса `POST /orders`, токен пользователя `Bearer ...`, чтобы сервер понимал что заказ сделал зарегистрированный пользователь (время обработки заказа должно быть в разы больше, если вы всё правильно сделали).

### **Сделает код лучше:**

- При загрузки страницы, лучше отображать лоадер, пока вы не убедитесь, что пользователь залогинен. А то сейчас вы сначала попадаете на страницу логина, и через пару секунд на страницу профиля.

#### **Хочу напомнить, что работа может быть принята, только после исправления всех "Надо исправить". Перед отправкой на ревью, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка F12).**

piladik committed:
Small fixes:
1. Добавил кастомный хук useForm
2. Поправил функцию updateUserRequest:
    - Теперь в ней используется функция refreshWithRequest
    - Теперь в ней создается объект с инпутами, вместо возвращения двух разных фанкций в зависимости от инпутов
3. В reducers.auth теперь user initialState = null
4. Поправил preloader. Теперь в reducers.auth есть флаг authChecked, который принимает значение true во время диспатча getUser в компоненте App и значение false после диспатча logout.

piladik committed:
Order request update
1. Теперь id булок есть в в начале и конце массива
2. Запрос /orders выполняется с header: {Authorization: accessToken}

ArslanMustafin reviewed (approved):
## Code Review

Здравствуйте!

Отличная работа.

### **Вы молодец! Работа принята, удачи в дальнейшем обучении!**

- Все замечания были исправлены;
